Add Dockerfile for development

Head0nF1re opened the pull request:
### What is the reason for this PR?

- Dockerfile to build an image with the FakerPHP dev dependencies. 

- It doesn't impact end users, it just improves the DX (Developer Experience) for anyone that wants to contribute to the repo without having to worry about installing any dependency (e.g. php, composer, make) on the host machine.


You just need to:
1. Build the image: `docker build -t php8.3-composer .`
2. Add an alias: `alias php8.3="docker run -it --rm -v .:/app -w /app php8.3-composer"`
3. Use it: `php8.3 make cs`

**Note**:
- The script that is added with `ADD` is meant to facilitate the installation of extensions. You can verify that the script is even mentioned in the [README of the offical PHP image](https://hub.docker.com/_/php):

> If you are having difficulty figuring out which Debian or Alpine packages need to be installed before docker-php-ext-install, then have a look at [the install-php-extensions project⁠](https://github.com/mlocati/docker-php-extension-installer). This script builds upon the docker-php-ext-* scripts and simplifies the installation of PHP extensions by automatically adding and removing Debian (apt) and Alpine (apk) packages.

- The image defaults to PHP version `8.3` since there seems to be a dependency that requires a max of 8.3.X. Composer defaults to `latest`. You can override these defaults using `docker build --build-args`, allowing you to have multiple versions to work with (by setting multiple alias).

Let me know what you think, I will add some info to the README if you find this useful.

Head0nF1re committed:
Add Dockerfile for development

- Dockerfile to build an image with the FakerPHP dev dependencies.
  Allows devs to work on the repo without having to install any
  dependencies (e.g. php, composer, make) on the host machine.

bram-pkg commented on Dockerfile:
```diff
@@ -0,0 +1,27 @@
+ARG PHP_VERSION=8.3 \
+	COMPOSER_VERSION=latest
+
+# At the moment, it's NOT possible to interpolate inside the COPY --from
```
Today I learned!

bram-pkg commented on Dockerfile:
```diff
@@ -0,0 +1,27 @@
+ENV COMPOSER_ROOT_VERSION=2.0.x-dev
+
+ARG USERNAME=faker \
+ 	USER_UID=1000 \
+	USER_GID=$USER_UID
```
Is this a mistake? Shouldn't the UID be variable to the arg

Head0nF1re replied:
What do you mean by:
> be variable to the arg

bram-pkg replied:
I meant that in the code snippet there you are binding the UID to the GID in the container. Is that on purpose?

Head0nF1re replied:
> I meant that in the code snippet there you are binding the UID to the GID in the container. Is that on purpose?

Yeah, it's common nowadays to create a user belonging to it's own group. Usually the first user created will have `gid` equal to `uid`. For example:

- https://wiki.archlinux.org/title/Users_and_groups#Example_adding_a_user
- https://code.visualstudio.com/remote/advancedcontainers/add-nonroot-user#_creating-a-nonroot-user

So, in this case, any file created will likely default to your host user and not have any file permissions issue between host and container.

bram-pkg commented on Dockerfile:
```diff
@@ -0,0 +1,27 @@
+		gzip \
+		unzip
+
+ADD --chmod=0755 https://github.com/mlocati/docker-php-extension-installer/releases/latest/download/install-php-extensions /usr/local/bin/
```
Could we use the same mechanic here as with composer: copying the script from the docker image

Head0nF1re replied:
Yeah, it would be fine. `ADD` needs to download the file each time to calculate the hash and decide if it should invalidate the cache, using the image would skip the downloading part which is better.

We just need to add some info in the README warning about potentially stale images and the need to use `build --pull` as needed. Could also use something like `watchtower` but since this is just for local development I don't see a need

Head0nF1re replied:
Removed the `ADD` and added the extension-installer image

pimjansen commented on Dockerfile:
```diff
@@ -0,0 +1,27 @@
+# https://getcomposer.org/doc/articles/troubleshooting.md#root-package-version-detection
+ENV COMPOSER_ROOT_VERSION=2.0.x-dev
+
+ARG USERNAME=faker \
```
I suggest using the organisation me as reference

Head0nF1re replied:
> I suggest using the organisation me as reference

Do you mean, using the organization name (`FakerPHP`) for the `USERNAME` or do you mean something related to `COMPOSER_ROOT_VERSION`?

Head0nF1re replied:
Changed `USERNAME` to FakerPHP. If it's only that, it's resolved

pimjansen commented on Dockerfile:
```diff
@@ -0,0 +1,27 @@
+    && useradd --uid $USER_UID --gid $USER_GID -m $USERNAME
+
+RUN apt-get update && apt-get install -y \
+		gzip \
```
For what is gzip needed?

Head0nF1re replied:
Good question lol. For some reason I thought I needed it for `composer` but after a clean `git clone` it seems only `unzip` is needed

Head0nF1re replied:
Removed gzip in the new commit

Head0nF1re committed:
Extension-installer image, gzip and USERNAME

- Add extension-install image instead of using ADD.
- Remove gzip since it's not needed.
- Change USERNAME to FakerPHP.

pimjansen commented:
Is it an idea to also integrate our tooling to run via docker instead? Since right now, what is a solid way to run stuff without complex cli commands?

Head0nF1re commented:
> Is it an idea to also integrate our tooling to run via docker instead

@pimjansen What do you mean by this?

<br/>


> what is a solid way to run stuff without complex cli commands

After building the image and adding an alias, the commands are pretty simple, e.g.: `php8.3 make cs`. 
You could also use `docker compose`, and run the commands inside the container `docker-compose exec foo bash`, not sure if it's worth it. Or, do you mean something else?

Also, do you want me to change the target branch of this PR to `1.24`?
